[BE]: Update cudnn to 9.1.0.70

eqy opened the pull request:
cuDNN has managed to upload cu11 and cu12 wheels for ~~9.0.0.312~~ 9.1.0.70, so trying this out...

CC @Skylion007 @malfet

cc @csarofeen @ptrblck @xwang233 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

eqy committed:
check in

Skylion007 reviewed (approved):
Looks good assuming you fix the Caffe2 issue.

eqy commented:
@Skylion007 do we anticipate #122527 landing in the near-future? caffe2 fixes might be involved and look something like #115719 which might be a bit involved if caffe2 is being removed soon

eqy committed:
bump to 9.1.0.70

johnnynunez commented:
9.1.1 is out

eqy commented:
@pytorchmergebot rebase (checking to see if caffe2 issue has been resolved)

eqy commented:
@pytorchmergebot rebase

Skylion007 commented:
@eqy Looks like the issues were fixed. Let's merge.

eqy commented:
@pytorchmergebot merge

Skylion007 commented:
@eqy Yeah looks like we need a builder PR: https://github.com/pytorch/builder/blob/ead2e21b24038ed9117060bb2c00bc79c4e91d85/CUDA_UPGRADE_GUIDE.MD?plain=1

Skylion007 commented:
@malfet You helped with the last cudnn PR, mind helping this land with the corresponding builder PR?

ptrblck commented:
CC @nWEIdia for the corresponding `pytorch/builder` PR.

nWEIdia commented:
[like]  Wei Wang reacted to your message:
________________________________
From: ptrblck ***@***.***>
Sent: Tuesday, May 14, 2024 5:02:43 PM
To: pytorch/pytorch ***@***.***>
Cc: Wei Wang ***@***.***>; Mention ***@***.***>
Subject: Re: [pytorch/pytorch] [BE]: Update cudnn to 9.1.0.70 (PR #123475)


CC @nWEIdia<https://github.com/nWEIdia> for the corresponding pytorch/builder PR.

—
Reply to this email directly, view it on GitHub<https://github.com/pytorch/pytorch/pull/123475#issuecomment-2110714895>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/BCHE4QEGJ37RY6HTHZ4Z5VDZCI7THAVCNFSM6AAAAABFZZ7ITOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDCMJQG4YTIOBZGU>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

Skylion007 commented:
The workflow lint checker also seems to be bugged out. ;-;

nWEIdia commented:
The builder branch has been created, to test this PR, I recall we may need to change the following file as well since cudnn gets bumped from 8 to 9, e.g. https://github.com/pytorch/pytorch/blob/updatecudnn9/.ci/docker/build.sh#L94 

We can test with companion builder branch via recommendations in https://github.com/pytorch/pytorch/pull/124014#issuecomment-2058144686 and https://github.com/pytorch/builder/pull/1822 
and fix lint and resolve conflicts (https://github.com/pytorch/pytorch/pull/125944 went in).

nWEIdia commented:
@atalman just reminded me that https://github.com/pytorch/pytorch/pull/121684/files generate_binary_build_matrix file may also need to be changed to reflect cudnn change.

nWEIdia commented:
@atalman and I would recommend this PR waiting for https://github.com/pytorch/pytorch/pull/121956, which is enabling CUDA 12.4 CI. Please wait for 12.4 CI to land first, thanks!

eqy committed:
update names

Skylion007 commented:
@nWEIdia CUDA 12.4 in, let's merge the NCCL and CUDNN dependency updates too.

bhack commented:
Can we link this to the ticket https://github.com/pytorch/pytorch/pull/123475 so that it will be autoclosed?

nWEIdia commented on .ci/docker/build.sh:
```diff
@@ -91,7 +91,7 @@ _UCC_COMMIT=20eae37090a4ce1b32bcce6144ccad0b49943e0b
-  pytorch-linux-focal-cuda12.4-cudnn8-py3-gcc9)
+  pytorch-linux-focal-cuda12.4-cudnn9-py3-gcc9)
     CUDA_VERSION=12.4.0
     CUDNN_VERSION=8
```
@eqy, CUDNN_VERSION=8 has to be changed to CUDNN_VERSION=9

nWEIdia commented on .ci/docker/build.sh:
```diff
@@ -105,7 +105,7 @@ case "$image" in
-  pytorch-linux-focal-cuda12.1-cudnn8-py3-gcc9)
+  pytorch-linux-focal-cuda12.1-cudnn9-py3-gcc9)
     CUDA_VERSION=12.1.1
     CUDNN_VERSION=8
```
Same here since you chnaged install_cudnn.sh for all cuda versions?

nWEIdia commented on .ci/docker/build.sh:
```diff
@@ -134,7 +134,7 @@ case "$image" in
-  pytorch-linux-focal-cuda12.1-cudnn8-py3-gcc9-inductor-benchmarks)
+  pytorch-linux-focal-cuda12.1-cudnn9-py3-gcc9-inductor-benchmarks)
     CUDA_VERSION=12.1.1
     CUDNN_VERSION=8
```
ditto

nWEIdia commented on .ci/docker/build.sh:
```diff
@@ -149,7 +149,7 @@ case "$image" in
-  pytorch-linux-focal-cuda12.1-cudnn8-py3.12-gcc9-inductor-benchmarks)
+  pytorch-linux-focal-cuda12.1-cudnn9-py3.12-gcc9-inductor-benchmarks)
     CUDA_VERSION=12.1.1
     CUDNN_VERSION=8
```
same

nWEIdia commented on .ci/docker/build.sh:
```diff
@@ -164,7 +164,7 @@ case "$image" in
-  pytorch-linux-focal-cuda11.8-cudnn8-py3-gcc9)
+  pytorch-linux-focal-cuda11.8-cudnn9-py3-gcc9)
     CUDA_VERSION=11.8.0
     CUDNN_VERSION=8
```
same

nWEIdia commented on .github/scripts/generate_binary_build_matrix.py:
12.4 also needs to be changed. Only changed 12.1 and 11.8

nWEIdia reviewed (changes requested):
cudnn9 has to show up in docker image, so the build.sh under .ci/docker needs to embrace cudnn v9 in full.

bhack commented:
> cudnn9 has to show up in docker image, so the build.sh under .ci/docker needs to embrace cudnn v9 in full.

`12.4.1` Images are with cudnn `9.1.0.70`
https://hub.docker.com/layers/nvidia/cuda/12.4.1-cudnn-runtime-ubuntu20.04/images/sha256-de9acc3e4d3aace101541899b95f6a4af897994124713f7131dffaf9967cb514?context=explore

But this is only eventually valid for pytorch nightly images.

In the CI Docker images I suppose we install cudnn `manually` https://github.com/pytorch/pytorch/issues/126005#issuecomment-2125841248

atalman commented:
Since we have generated docker images with cudnn 9.0. Please temporarily pin docker manywheel builder images to these values:
```
pytorch/manylinux-builder:cuda12.1-7e4bdab63653ad6af8d92fdfa69d7e96780ce258
pytorch/manylinux-builder:cuda11.8-7e4bdab63653ad6af8d92fdfa69d7e96780ce258
pytorch/manylinux-builder:cuda12.4-7e4bdab63653ad6af8d92fdfa69d7e96780ce258
```

Skylion007 commented:
Ugh, cudnn 9.1.1.17 just had pip wheels generated too. :(

nWEIdia commented on .ci/docker/common/install_cudnn.sh:
```diff
@@ -5,19 +5,14 @@ if [[ ${CUDNN_VERSION} == 8 ]]; then
-    elif [[ ${CUDA_VERSION:0:4} == "11.8" ]]; then
-        CUDNN_NAME="cudnn-linux-x86_64-8.7.0.84_cuda11-archive"
-        curl --retry 3 -OLs https://developer.download.nvidia.com/compute/redist/cudnn/v8.7.0/local_installers/11.8/${CUDNN_NAME}.tar.xz
+        CUDNN_NAME="cudnn-linux-x86_64-9.1.0.70_cuda11-archive"
```
cuda11 does not look right. 
We still support cuda 11.8, right? So three versions: cuda 12.1, cuda 12.4, and cuda 11.8

eqy committed:
fixes

eqy committed:
more fixes

eqy committed:
fix

eqy committed:
change image

eqy committed:
try using CUDNN_VERSION==8 specialized name

eqy committed:
more version fixes

eqy committed:
sed

nWEIdia commented:
there are a few places like: https://github.com/pytorch/pytorch/blob/updatecudnn9/.github/workflows/trunk.yml#L91 
https://github.com/pytorch/pytorch/blob/updatecudnn9/.github/workflows/trunk.yml#L115 
https://github.com/pytorch/pytorch/blob/updatecudnn9/.github/workflows/trunk.yml#L129 

Please check the other yml files too.

nWEIdia reviewed (approved):
Epic changes!

atalman commented on .github/scripts/generate_binary_build_matrix.py:
```diff
@@ -42,7 +42,7 @@
         "nvidia-cuda-runtime-cu11==11.8.89; platform_system == 'Linux' and platform_machine == 'x86_64' | "
         "nvidia-cuda-cupti-cu11==11.8.87; platform_system == 'Linux' and platform_machine == 'x86_64' | "
-        "nvidia-cudnn-cu11==8.7.0.84; platform_system == 'Linux' and platform_machine == 'x86_64' | "
+        "nvidia-cudnn-cu11==9.1.0.70; platform_system == 'Linux' and platform_machine == 'x86_64' | "
```
```suggestion
        "nvidia-cudnn-cu12==9.1.0.70; platform_system == 'Linux' and platform_machine == 'x86_64' | "
```

atalman commented on .github/scripts/generate_binary_build_matrix.py:
```diff
@@ -42,7 +42,7 @@
         "nvidia-cuda-runtime-cu11==11.8.89; platform_system == 'Linux' and platform_machine == 'x86_64' | "
         "nvidia-cuda-cupti-cu11==11.8.87; platform_system == 'Linux' and platform_machine == 'x86_64' | "
-        "nvidia-cudnn-cu11==8.7.0.84; platform_system == 'Linux' and platform_machine == 'x86_64' | "
+        "nvidia-cudnn-cu12==9.1.0.70; platform_system == 'Linux' and platform_machine == 'x86_64' | "
```
```suggestion
        "nvidia-cudnn-cu11==9.1.0.70; platform_system == 'Linux' and platform_machine == 'x86_64' | "
```

atalman replied:
put back cu11

Skylion007 commented:
It would be nice if we could update to 9.1.1.17 as it appears as though it has fixed some bad heuristics in 9.1.0 that could cause perf regressions.

nWEIdia committed:
Update CUDA_ARCHES_CUDNN_VERSION

eqy committed:
oops remove hardcoded cudnn8 condition

eqy commented:
> It would be nice if we could update to 9.1.1.17 as it appears as though it has fixed some bad heuristics in 9.1.0 that could cause perf regressions.
> 
> > On NVIDIA Hopper and Ampere architectures, fused flash attention bprop would return an execution failed when sequence lengths greater than or equal to 512k were used. This issue has been fixed.
> > Fixed heuristics regression in version 9.1.0 with mixed input precision MatMul/convolution use cases.
> > On NVIDIA Hopper architectures, fused flash attention bprop could lead to incorrect results in version 8.9.7; we missed documenting the issue in that version. This issue was fixed in version 9.0.0.
> > The scale and bias tensor descriptor checks for Layer and RMS Normalization have been updated to require the normalizing dimensions to be specified in the correct order. These checks were missing in cuDNN version 9.0 and prior version
> > https://docs.nvidia.com/deeplearning/cudnn/latest/release-notes.html

We're a bit hesistant to move to 9.1.1 here last minute as 9.1.0 has been much more thoroughly tested in NGC containers while 9.1.1 has had basically no testing in that form. AFAIK mixed-precision conv/matmul are a rare use-case that isn't supported at the moment? e.g., even for batchnorm it has not been merged yet https://github.com/pytorch/pytorch/pull/121124

Layer and RMSNorm do not go through cuDNN; see also https://github.com/pytorch/pytorch/pull/125350

eqy committed:
oops

eqy committed:
in case

nWEIdia reviewed (changes requested):
Line 131 in workflows/generated-linux-binary-manywheel-nightly.yml is wrong, 
causes https://github.com/pytorch/pytorch/actions/runs/9325276243/job/25674005681#step:17:450 to download wrong cudnn version. 

Note that the generated-* file should not be manually changed and needs to be generated from .github/regenerate.sh 

I will push a fix.

nWEIdia committed:
Fix test failures caused by downloading wrong cudnn version.
Run ./github/regenerate.sh after modifying
scripts/generate_binary_build_matrix.py

nWEIdia commented:
for future reference, the failures in https://github.com/pytorch/pytorch/actions/runs/9327166331/job/25676708999 means this PR was testing with pytorch/builder main branch and are missing changes in https://github.com/pytorch/builder/pull/1847 (manywheel/build_cuda.sh changes, to be specific). This is also the reason to land these two PRs together. 

```
2024-06-01T05:14:23.4785350Z ++ for filepath in '"${DEPS_LIST[@]}"'
2024-06-01T05:14:23.4786766Z +++ basename /usr/local/cuda/lib64/libcudnn_adv_infer.so.8
2024-06-01T05:14:23.4794447Z ++ filename=libcudnn_adv_infer.so.8
2024-06-01T05:14:23.4794921Z ++ destpath=libtorch/lib/libcudnn_adv_infer.so.8
2024-06-01T05:14:23.4795861Z ++ [[ /usr/local/cuda/lib64/libcudnn_adv_infer.so.8 != \l\i\b\t\o\r\c\h\/\l\i\b\/\l\i\b\c\u\d\n\n\_\a\d\v\_\i\n\f\e\r\.\s\o\.\8 ]]
2024-06-01T05:14:23.4797021Z ++ cp /usr/local/cuda/lib64/libcudnn_adv_infer.so.8 libtorch/lib/libcudnn_adv_infer.so.8
2024-06-01T05:14:23.4804470Z cp: cannot stat ‘/usr/local/cuda/lib64/libcudnn_adv_infer.so.8’: No such file or directory
```

eqy commented:
might be related to rocm failures? https://github.com/pytorch/pytorch/pull/125396#issuecomment-2142646195

nWEIdia commented:
> might be related to rocm failures? [#125396 (comment)](https://github.com/pytorch/pytorch/pull/125396#issuecomment-2142646195)

Yeah, we may need to rebase to avoid this failure because the rocm change was reverted but this PR probably did not include the revert.

nWEIdia commented:
linux conda build jobs are failing for this PR, 
but linux conda build jobs are not failing for the equivalent PR + pytorch/builder change (pytorch/pytorch #127589), again meaning this PR depends on pytorch/builder PR. 

Failure (this PR): https://github.com/pytorch/pytorch/actions/runs/9327166376/job/25676701472
Working (equivalent PR #127589 ): https://github.com/pytorch/pytorch/actions/runs/9327183348/job/25676738895 

The difference between failure and working is again the libcudnn_adv_infer.so.8 -> libcudnn_adv.so.9 difference (working)

nWEIdia commented:
Both this PR and the other PR fails for libtorch build but for difference reasons: 
1) this PR fails  https://github.com/pytorch/pytorch/actions/runs/9327166331/job/25676708999, because again 

2024-06-01T05:14:23.4786766Z +++ basename /usr/local/cuda/lib64/libcudnn_adv_infer.so.8
2024-06-01T05:14:23.4794447Z ++ filename=libcudnn_adv_infer.so.8
2024-06-01T05:14:23.4794921Z ++ destpath=libtorch/lib/libcudnn_adv_infer.so.8
2024-06-01T05:14:23.4795861Z ++ [[ /usr/local/cuda/lib64/libcudnn_adv_infer.so.8 != \l\i\b\t\o\r\c\h\/\l\i\b\/\l\i\b\c\u\d\n\n\_\a\d\v\_\i\n\f\e\r\.\s\o\.\8 ]]
2024-06-01T05:14:23.4797021Z ++ cp /usr/local/cuda/lib64/libcudnn_adv_infer.so.8 libtorch/lib/libcudnn_adv_infer.so.8
2024-06-01T05:14:23.4804470Z cp: cannot stat ‘/usr/local/cuda/lib64/libcudnn_adv_infer.so.8’: No such file or directory

steering this PR towards pytorch/builder branch should fix 

2) the other PR failed here https://github.com/pytorch/pytorch/actions/runs/9327183345/job/25676742766 (that is linux-binary-libtorch-cxx11-abi  job) but successful here (https://github.com/pytorch/pytorch/actions/runs/9327183368/job/25679146488), the reason is the docker image has to use the customized one (not cu124-main, cu121-main, cu118-main, but the one with the {sha}. One succeed and one failure means, this PR needs to modify the .github/scripts/generate_binary_build_matrix.py to be consistent for both libtorch builds.

nWEIdia commented:
In addition, https://github.com/pytorch/pytorch/actions/runs/9327166354/job/25676707213#step:16:4504 is saying arm cuda was not picking up cudnn v9 changes.

nWEIdia committed:
Fix errors in #127589 (libtorch build)

atalman committed:
Temporary change to be removed before merging

nWEIdia reviewed (approved):
It seems only 1) rebase and 2) revert testing changes are needed (to be done last minute and separate from the rebase). 
Need to watch out inductor tiny* model accuracy failures.

nWEIdia commented:
@pytorchbot rebase

tinglvv commented:
cuda-aarch64 builds are failing due to 
`FileNotFoundError: [Errno 2] No such file or directory: '/usr/local/cuda/lib64/libcudnn.so.9'`
https://github.com/pytorch/pytorch/actions/runs/9333823315/job/25691260378?pr=123475

We would need to merge https://github.com/pytorch/builder/pull/1847 which copies this file to ARM wheel.

eqy committed:
revert image specific change

eqy committed:
update expected failures

eqy committed:
undo temp changes

eqy committed:
fix

Skylion007 commented:
Remaining failures seem unrelated. Let's coordinate the builder merge @malfet @albanD

atalman commented:
@pytorchmergebot merge -f "All required workflows are green"

eqy commented:
For inductor EfficientConvBN failure, see #127984

huydhn commented:
@pytorchbot revert -m 'CUDA trunk jobs are pretty red after this change, and the forward fix https://github.com/pytorch/pytorch/pull/127984 does not look working' -c weird

atalman commented:
@huydhn Here is the signal on the same PR before merging it: https://github.com/pytorch/pytorch/actions/runs/9358133192/job/25760440684
This looks like land race.

nWEIdia commented:
> @huydhn Here is the signal on the same PR before merging it: https://github.com/pytorch/pytorch/actions/runs/9358133192/job/25760440684 This looks like land race.

Agree. There should only be https://github.com/pytorch/pytorch/issues/127990 after merging cudnn v9. We just missed one model beit which we could have communicated. 

![image](https://github.com/pytorch/pytorch/assets/143543872/91f35730-d22b-450b-bc6b-993fd954bda2)

Just that one model.

nWEIdia committed:
Adjust cu121 beit accuracy expectation.
Testing cudnn v9 changes with pinned docker image and builder PR
Arm CUDA will unfortunately still fail due to pinned buggy(for arm only, x86_64 ok) docker image

eqy commented:
@huydhn looks like another fix of the same type is needed for grads https://github.com/pytorch/pytorch/pull/123475, I'll open

nWEIdia commented:
For this PR, after https://github.com/pytorch/pytorch/pull/127825, we need to rebase **AND** cudnn8-> cudnn9 related changes.

nWEIdia committed:
Conda docker images were generated using a different main commit of
pytorch/builder

nWEIdia committed:
Sorry, Forgot to run ./regenerate.sh

nWEIdia committed:
Change cudnn8 -> cudnn9 after cuda 12.4 move

nWEIdia committed:
Fix libtorch cxx11 docker image pin

nWEIdia committed:
Revert "Update test_efficient_conv_bn_eval.py"

This reverts commit 86172c533c56227eea0c8fce46443e0d1b386ab2.

nWEIdia committed:
Roll back temp changes and cherrypicks

atalman commented:
@pytorchmergebot merge -f "lint is green"
